numpyro/infer/einstein/kernels.py

```diff
@@ -0,0 +1,388 @@
+    :math: `k(x,y) = \\sum_i w_ik_i(x,y)`
+
+    ** Reference **
+    1. *Stein Variational Gradient Descent as Moment Matching* by Liu and Wang
+
+    :param ws: Weight of each kernel in the mixture
+    :param kernel_fns: Different kernel functions to mix together
+    """
+
+    def __init__(self,  ws: List[float], kernel_fns: List[SteinKernel], mode='norm',):
+        assert len(ws) == len(kernel_fns)
+        assert len(kernel_fns) > 1
+        assert all(kf.mode == mode for kf in kernel_fns)
+        self.ws = ws
+        self.kernel_fns = kernel_fns
+
+    @property
+    def mode(self):
+        return self.kernel_fns[0].mode
+
+    def compute(self, particles, particle_info, loss_fn):
+        kernels = [kf.compute(particles, particle_info, loss_fn) for kf in self.kernel_fns]
+
+        def kernel(x, y):
+            res = self.ws[0] * kernels[0](x, y)
+            for w, k in zip(self.ws[1:], kernels[1:]):
+                res = res + w * k(x, y)
+            return res
+
+        return kernel
+
+
+class HessianPrecondMatrix(PrecondMatrix):
+    """
+    Calculates the constant precondition matrix based on the negative Hessian of the loss from [1].
+
+    ** References: **
+    1. *Stein Variational Gradient Descent with Matrix-Valued Kernels* by Wang, Tang, Bajaj and Liu
+    """
+
+    def compute(self, particles, loss_fn):
+        hessian = -jax.vmap(jax.hessian(loss_fn))(particles)
+        return hessian
+
+
+class PrecondMatrixKernel(SteinKernel):
+    """
+    Calculates the const preconditioned kernel
+    :math: `k(x,y) = Q^{-\\frac{1}{2}}k(Q^{\\frac{1}{2}}x, Q^{\\frac{1}{2}}y)Q^{-\\frac{1}{2}},`
+    or anchor point preconditioned kernel
+    :math: `k(x,y) = \\sum_{l=1}^m k_{Q_l}(x,y)w_l(x)w_l(y)`
+    both from [1].
+
+    ** References: **
+    1. "Stein Variational Gradient Descent with Matrix-Valued Kernels" by Wang, Tang, Bajaj and Liu
+
+    :param precond_matrix_fn: The constant preconditioning matrix
+    :param inner_kernel_fn: The inner kernel function
+    :param precond_mode: How to use the precondition matrix, either constant ('const')
+                         or as mixture with anchor points ('anchor_points')
+    """
+
+    def __init__(self, precond_matrix_fn: PrecondMatrix, inner_kernel_fn: SteinKernel,
+                 precond_mode='anchor_points'):
+        assert inner_kernel_fn.mode == 'matrix'
+        assert precond_mode == 'const' or precond_mode == 'anchor_points'
+        self.precond_matrix_fn = precond_matrix_fn
+        self.inner_kernel_fn = inner_kernel_fn
+        self.precond_mode = precond_mode
+
+    @property
+    def mode(self):
+        return 'matrix'
+
+    def compute(self, particles, particle_info, loss_fn):
+        qs = self.precond_matrix_fn.compute(particles, loss_fn)
+        if self.precond_mode == 'const':
+            qs = jnp.expand_dims(jnp.mean(qs, axis=0), axis=0)
+        qs_inv = jnp.linalg.inv(qs)
+        qs_sqrt = sqrth(qs)
+        qs_inv_sqrt = sqrth(qs_inv)
```

Comment:
it would be faster to have a function that returns 3 terms: `qs_inv_scale_tril, qs_sqrt, qs_inv_postdef` so we only need to use eig decomposition 1 time: something like
```
def ...(m):
    mlambda, mvec = jnp.linalg.eigh(m)
    ...  # address precision issue of mlambda
    mlambdasqrt = mlambda ** 0.5
    mvec_t = jnp.swapaxes(mvec, -2, -1)
    qs_sqrt = mvec @ mlambdasqrt @ mvec_t
    qs_inv_sqrt = mvec @ (1 / mlambdasqrt) @ mvec_t
    qs_inv_posdef = mvec @ (1 / mlambda) @ mvec_t
    return qs_sqrt, qs_inv_sqrt, qs_inv_posdef
```

In case `x` is a vector, you can replace `A @ diag(x)` with `A * x`, which should be much faster. Using this fact, you can reduce computing time for `mvec @ mlambdasqrt` and other terms in case `mlambda.ndim == 1`.